experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1192 @@
+                    state = PropertyValueBegin;
+                }
+                // parse the property value
+                (PropertyValue | PropertyValueBegin, c) if c != end => {
+                    value_buffer.push(c);
+                    self.iter.next();
+                    state = PropertyValue;
+                }
+                (_, c) => return self.error_here(PEK::UnexpectedChar(c)),
+            }
+        }
+
+        if !equality {
+            self.inverted = !self.inverted;
+        }
+
+        let (set, inverted) = self
+            .load_property_codepoints(&key_buffer, &value_buffer)
+            // any error that does not already have an offset should use the appropriate property offset
+            .map_err(|e| e.or_with_offset(property_offset))?;
+        if inverted {
+            self.inverted = !self.inverted;
+        }
+        self.single_set.add_set(&set.to_code_point_inversion_list());
+
+        Ok(())
+    }
+
+    // also returns whether the set needs to be inverted or not
+    fn load_property_codepoints(&self, key: &str, value: &str) -> Result<(CodePointSetData, bool)> {
+        // we support:
+        // [:gc = value:]
+        // [:sc = value:]
+        // [:scx = value:] -- TODO: implement scx
+        // [:value:] - looks up value in gc, sc and scx
+        // [:prop:] - binary property, returns codepoints that have the property
+        // [:prop = truthy/falsy:] - same as above
+
+        let mut inverted = false;
+
+        if value.is_empty() {
+            // key is binary property, or a value of gc, sc, scx
+
+            // try loading a binary property, if it fails, try loading a gc, sc, scx value
+            let set = load_for_ecma262_unstable(self.property_provider, key)
+                .or_else(|_| self.try_load_general_category(key))
+                .or_else(|_| self.try_load_script(key))?;
+            // .or_else(|_| self.try_load_script_extensions(key))?;
+
+            Ok((set, inverted))
+        } else {
+            // key is gc, sc, scx
+            // value is a property value
+            // OR
+            // key is a binary property and value is a truthy/falsy value
+
+            // UnicodeProperty::parse_ecma262_name would be good to have to avoid this duplication:
+            let set = match key {
+                "General_Category" | "gc" => self.try_load_general_category(value)?,
+                "Script" | "sc" => self.try_load_script(value)?,
+                // "Script_Extensions" | "scx" => {
+                //     // https://www.unicode.org/reports/tr24/#Script_Extensions
+                //     // the UnicodeSet utils page takes lists of scripts as values, but I'm not sure how to implement that using load_script_with_extensions_unstable.
+                //     // skipped for now.
+                _ => {
+                    // try prop = truthy/falsy case
+                    let set = load_for_ecma262_unstable(self.property_provider, key)
```

Comment:
Suggestion: this invocation of `load_for_ecma262_unstable` should go after the `if truthy == falsy` because you can exit early without hitting the data provider.

Reply:
Good point!